clippy_lints/src/methods/owned_to_owned.rs

```diff
@@ -0,0 +1,43 @@
+use crate::utils::{is_copy, match_def_path, match_type, span_lint_and_sugg, sugg};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::ExprKind;
+use rustc_lint::LateContext;
+use rustc_middle::ty::TyS;
+
+use super::OWNED_TO_OWNED;
+
+pub fn lint(cx: &LateContext<'_>, expr: &hir::Expr<'_>, expected_path: &[&str], expected_type: Option<&[&str]>) {
+    if_chain! {
+        if let ExprKind::MethodCall(method_path, _, [arg], _) = &expr.kind;
+        if let Some(meth_did) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if let Some(snippet) = sugg::Sugg::hir_opt(cx, arg);
+        let return_type = cx.typeck_results().expr_ty(&expr);
+        let input_type = cx.typeck_results().expr_ty(arg);
+        if TyS::same_type(return_type, input_type);
+        if match_def_path(cx, meth_did, expected_path) &&
+            expected_type.map_or(true, |expected_type_path| match_type(cx,input_type,expected_type_path));
+        then {
+            if is_copy(cx,return_type) {
+                // if return_type is copy, have the suggestion be to replace the call with the variable itself
+                //  this prevents fixing this lint, only to have clone_on_copy complain next.
+                span_lint_and_sugg(
+                    cx,OWNED_TO_OWNED,expr.span,
+                    &format!("using `{}` on an already owned type",method_path.ident.name),
+                    "replace this with",
+                    snippet.to_string(),
+                    Applicability::MaybeIncorrect
+                );
+            } else {
+                span_lint_and_sugg(
+                    cx,OWNED_TO_OWNED,method_path.ident.span,
+                    &format!("using `{}` on an already owned type",method_path.ident.name),
```

Comment:
"Using `to_vec` on a `Vec`" would be more clear.

```suggestion
                    &format!("using `{}` on a `{}`", method_path.ident.name, ty_name),
```

where `ty_name` is `cx.tcx.item_name(input_type.ty_adt_def().unwrap().did)`

Also (very picky nit) please remember spaces after commas inside `if_chain!`.

Reply:
I was actually looking for how to do that. Done.

Re the nit: Fixed and I will try to be more aware of the missing spaces.